File: python/commands.py
def encode_command(command, value):
    res = command
    res += ":"
    if not value:
        res += "null"
    else:
        res += str(value)
    return res


def decode_command(command):
    parts = command.split(":")
    if len(parts) != 2:
        raise ValueError("Invalid command format")
    return parts[0], parts[1].strip()


def decode_response(response):
    parts = response.split(":")
    if len(parts) != 3:
        raise ValueError("Invalid response format")
    if parts[1] != "ok":
        raise ValueError("Response indicates failure")
    return parts[2].strip()


def encode_response(command, status, value):
    res = command
    res += ":"
    res += status
    res += ":"
    if not value:
        res += "null"
    else:
        res += str(value)
    return res

File: commands/commands.go
package commands

import (
	"fmt"
	"strings"
)

func EncodeCommand(command string, value []byte) []byte {
	var sb strings.Builder
	sb.WriteString(command)
	sb.WriteString(":")
	if len(value) == 0 {
		sb.WriteString("null")
	} else {
		sb.WriteString(string(value))
	}
	sb.WriteString("\n")
	return []byte(sb.String())
}

func DecodeCommand(data []byte) (string, string, error) {
	parts := strings.SplitN(string(data), ":", 2)
	if len(parts) != 2 {
		return "", "", fmt.Errorf("invalid command, expected command:value, got %v", string(data))
	}
	return parts[0], strings.TrimSpace(parts[1]), nil
}

func DecodeResponse(res []byte) (string, error) {
	parts := strings.SplitN(string(res), ":", 3)
	if len(parts) != 3 {
		return "", fmt.Errorf("invalid response, expected command:status:result, got %v", string(res))
	}
	if parts[1] != "ok" {
		return "", fmt.Errorf("command failed with status: %s, result: %s", parts[1], parts[2])
	}
	return strings.TrimSpace(parts[2]), nil
}

func EncodeResponse(command string, status string, result []byte) []byte {
	var sb strings.Builder
	sb.WriteString(command)
	sb.WriteString(":")
	sb.WriteString(status)
	sb.WriteString(":")
	if len(result) == 0 {
		sb.WriteString("null")
	} else {
		sb.WriteString(string(result))
	}
	sb.WriteString("\n")
	return []byte(sb.String())
}
